test/features/build_server/compiler_interface_test.dart

```diff
@@ -0,0 +1,319 @@
+  List<SourceFile> files,
+  Map<String, String> fileMap,
+  CompilerConfig config,
+  ProjectCache cache, {
+  bool skipUpdateFiles = false,
+}) async {
+  final Completer<List<CompileStatusMessage>> completer = Completer();
+  final List<CompileStatusMessage> messageCache = [];
+
+  interface.onStateUpdate = (CompileStatusMessage m) {
+    messageCache.add(m);
+    if (m.status == CompileStatus.done) {
+      completer.complete(messageCache);
+    }
+    if (m.status == CompileStatus.failed) {
+      completer.completeError(StateError(m.payload));
+    }
+  };
+
+  await interface.init(fileMap, config, cache);
+
+  if (!skipUpdateFiles) {
+    final requiredFiles = await interface.sync(fileMap, config);
+
+    for (final file in requiredFiles) {
+      final srcFile = files.singleWhere((element) => element.id == file);
+      await interface.update(srcFile);
+    }
+  }
+
+  await interface.build();
+
+  return await completer.future;
+}
+
+Future<void> main() async {
+  MainInterface? interface;
+  ProjectCache? cache;
+
+  final config1 = CompilerConfig(
+    compiler: InstalledCompiler.gpp,
+    threads: 4,
+    sourceDirectories: ['./example'],
+    compilerFlags: ['-O2', '-DPRINTHELLO'],
+    linkerFlags: ['-flto'],
+  );
+
+  final files = await searchDirectory(Directory('./example'), (e) {
+    final isSource = FileExtension.sources.contains(p.extension(e.path));
+    final isHeader = FileExtension.headers.contains(p.extension(e.path));
+
+    if (e is File && (isSource || isHeader)) {
+      return SourceFile.fromFile(0, e, entryDirectory: Directory('./example'));
+    }
+
+    return null;
+  });
+
+  final fileMap = {
+    for (final file in files) file.id: await file.hash,
+  };
+
+  setUp(() async {
+    await clearTestCacheDirectory(id: 'compiler_interface_test');
+
+    interface = await MainInterface.launch();
+    cache = await createDummyProjectCache(id: 'compiler_interface_test');
+  });
+
+  tearDown(() async => await clearTestCacheDirectory(id: 'compiler_interface_test'));
+
+  test('Requires "init" call before answering other requests', () async {
+    await expectLater(
+      interface!.hasCapacity(),
+      throwsStateError,
+    );
+
+    await interface!.init(fileMap, config1, cache);
+
+    await expectLater(
+      interface!.hasCapacity(),
+      completion(equals(true)),
+    );
+  });
+
+  test('Can sychronize and update a project', () async {
+    await expectLater(() async => await interface!.init(fileMap, config1, cache), returnsNormally);
+
+    final requiredFiles = await interface!.sync(fileMap, config1);
+
+    expect(requiredFiles, unorderedEquals(fileMap.keys));
+
+    for (final file in requiredFiles) {
+      final srcFile = files.singleWhere((element) => element.id == file);
+      await expectLater(
+        () async => await interface!.update(srcFile),
+        returnsNormally,
+      );
+    }
+
+    await expectLater(() async => await interface!.build(), returnsNormally);
+  });
+
```

Comment:
I would like a test case where you test if already synched files are not requested again and changed files are requested again

Reply:
That would be repeated and is already covered in compiler or cache test. This test is for the interface not the underlying compiler

Reply:
So you think, that the Interface only forwards and does not manipulate the data in a way that would be worth to test?

Reply:
It doesn't change the information. It just changes some format like the when `didFinishCompilation` is called it sends a `CompileStatusMessage`. The part u asked about is just a `ProjectUpdateResponse` wrapped around the results of `Compiler.update` which is tested.

Reply:
Please resolve if you can verify my statement and don't have further questions. Otherwise I'd be happy to answer 'em all

Reply:
Please have a look at our code coverage of the related files. There simply is nothing we could possibly test without writing a test for the already tested underlying logic. If something is changed in `CompilerInterface` we'd need to test it anyway. And if something changes in the underlying parts it's already covered. I think unit tests should test one specific unit and not some other (related) unit

Reply:
I get what you are saying, but we also test the `WebServer` despite already having tests for the `Session`'s. It still passes through some other logic, which might change in the future. I really want this PR to land soon, so I'll give you the choice to either resolve this change request, or write the little test. Either way, I think we both stated our views clearly. You are mostly responsible for the compilation interface, so you may decide.

Reply:
Thank you very much, for letting me decide. As soon as there is some logic, there will be tests